packet.go

```diff
@@ -0,0 +1,217 @@
+// Checksum computations based on logic in github.com/google/gopacket
+//
+// Copyright 2012 Google, Inc. All rights reserved.
+// Copyright 2009-2011 Andreas Krennmair. All rights reserved.
+
+package gonat
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/getlantern/errors"
+)
+
+// TCPFlags are the different flags supported in the TCP header
+const (
+	TCPFlagRST = 0x04
+)
+
+var (
+	networkByteOrder = binary.BigEndian
+)
+
+func parseIPPacket(raw []byte) (*IPPacket, error) {
+	ipVersion := uint8(raw[0]) >> 4
+	if ipVersion != 4 {
+		return nil, errors.New("Unsupported ip protocol version: %v", ipVersion)
+	}
+
+	pkt := &IPPacket{Raw: raw, IPVersion: ipVersion}
+	return pkt.parseV4()
+}
+
+type IPPacket struct {
+	Raw       []byte
+	IPVersion uint8
+	IPProto   uint8
+	SrcAddr   *net.IPAddr
+	DstAddr   *net.IPAddr
+	Header    []byte
+	Payload   []byte
+}
+
+func (pkt *IPPacket) parseV4() (*IPPacket, error) {
+	ihl := uint8(pkt.Raw[0]) & 0x0F
+	length := networkByteOrder.Uint16(pkt.Raw[2:4])
+	if length < 20 {
+		return pkt, errors.New("Invalid (too small) IP length (%d < 20)", length)
+	} else if ihl < 5 {
+		return pkt, errors.New("Invalid (too small) IP header length (%d < 5)", ihl)
+	} else if int(ihl*4) > int(length) {
+		return pkt, errors.New("Invalid IP header length > IP length (%d > %d)", ihl, length)
+	} else if int(ihl)*4 > len(pkt.Raw) {
+		return pkt, errors.New("Not all IP header bytes available")
+	}
+
+	pkt.Header = pkt.Raw[:ihl*4]
+	pkt.Payload = pkt.Raw[ihl*4:]
+	pkt.IPProto = uint8(pkt.Header[9])
+	pkt.SrcAddr = &net.IPAddr{IP: net.IP(pkt.Header[12:16])}
+	pkt.DstAddr = &net.IPAddr{IP: net.IP(pkt.Header[16:20])}
+
+	return pkt, nil
+}
+
+// HasTCPFlag returns true if the packet is a TCP packet that has the given flag set.
+func (pkt *IPPacket) HasTCPFlag(flag uint8) bool {
```

Comment:
Why export this?

Reply:
Same reason as exporting the RST flag.

Reply:
Got it.  Sorry, I meant to mark this resolved.  I hadn't seen your response  to the RST flag comment when I left this.